File: scripts/prepare-npm-packages.ts
#!/usr/bin/env node
import { spawnSync } from 'node:child_process';
import { chmodSync, existsSync, mkdirSync, readFileSync, readdirSync, rmSync, statSync, writeFileSync } from 'node:fs';
import { basename, join } from 'node:path';
import process from 'node:process';

type TargetMetadata = {
  package: string;
  os: string;
  cpu: string;
};

type Arguments = {
  version?: string;
  artifactsDir?: string;
  outDir: string;
  dryRunPlaceholders: boolean;
  packDryRun: boolean;
};

const PACKAGE = '@casualjim/heimdall-sandbox';
const BINARY = 'heimdall-sandbox';
const TARGETS: Record<string, TargetMetadata> = {
  'x86_64-unknown-linux-gnu': {
    package: '@casualjim/heimdall-sandbox-linux-x64',
    os: 'linux',
    cpu: 'x64',
  },
  'aarch64-unknown-linux-gnu': {
    package: '@casualjim/heimdall-sandbox-linux-arm64',
    os: 'linux',
    cpu: 'arm64',
  },
  'aarch64-apple-darwin': {
    package: '@casualjim/heimdall-sandbox-darwin-arm64',
    os: 'darwin',
    cpu: 'arm64',
  },
};

function parseArguments(argv: string[]): Arguments {
  const args: Arguments = {
    outDir: 'target/npm-packages',
    dryRunPlaceholders: false,
    packDryRun: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    switch (arg) {
      case '--version':
        args.version = requireValue(argv, ++index, arg);
        break;
      case '--artifacts-dir':
        args.artifactsDir = requireValue(argv, ++index, arg);
        break;
      case '--out-dir':
        args.outDir = requireValue(argv, ++index, arg);
        break;
      case '--dry-run-placeholders':
        args.dryRunPlaceholders = true;
        break;
      case '--pack-dry-run':
        args.packDryRun = true;
        break;
      default:
        throw new Error(`unknown argument: ${arg}`);
    }
  }

  return args;
}

function requireValue(argv: string[], index: number, flag: string): string {
  const value = argv[index];
  if (!value || value.startsWith('--')) {
    throw new Error(`${flag} requires a value`);
  }
  return value;
}

function workspaceVersion(): string {
  const manifest = readText('Cargo.toml');
  let inWorkspacePackage = false;
  for (const line of manifest.split('\n')) {
    const trimmed = line.trim();
    if (trimmed === '[workspace.package]') {
      inWorkspacePackage = true;
      continue;
    }
    if (trimmed.startsWith('[') && trimmed !== '[workspace.package]') {
      inWorkspacePackage = false;
    }
    if (inWorkspacePackage) {
      const match = /^version\s*=\s*"([^"]+)"/.exec(trimmed);
      if (match) {
        return match[1];
      }
    }
  }
  throw new Error('workspace.package.version not found in Cargo.toml');
}

function readText(path: string): string {
  return readFileSync(path, 'utf8');
}

function writeJson(path: string, data: unknown): void {
  writeFileSync(path, `${JSON.stringify(data, null, 2)}\n`, 'utf8');
}

function writeExecutable(path: string, content: string | Buffer): void {
  writeFileSync(path, content);
  const mode = statSync(path).mode;
  chmodExecutable(path, mode);
}

function chmodExecutable(path: string, mode = statSync(path).mode): void {
  chmodSync(path, mode | 0o111);
}

function findArchive(artifactsDir: string, target: string): string {
  const prefix = `${BINARY}-${target}.tar.`;
  const matches = readdirSync(artifactsDir)
    .filter((entry) => entry.startsWith(prefix))
    .sort();
  if (matches.length === 0) {
    throw new Error(`missing cargo-dist archive for ${target} in ${artifactsDir}`);
  }
  return join(artifactsDir, matches[0]);
}

function extractBinary(archive: string, destination: string): void {
  const list = spawnChecked('tar', ['-tf', archive]).stdout.trim().split('\n');
  const member = list.find((entry) => basename(entry) === BINARY);
  if (!member) {
    throw new Error(`archive ${archive} does not contain ${BINARY}`);
  }
  const binary = spawnChecked('tar', ['-xOf', archive, member], 'buffer').stdout;
  writeExecutable(destination, binary);
}

function packageSlug(packageName: string): string {
  return packageName.replace(/^@casualjim\//, '');
}

function createPlatformPackage(outDir: string, target: string, meta: TargetMetadata, version: string, artifactsDir?: string): string {
  const packageDir = join(outDir, packageSlug(meta.package));
  const binDir = join(packageDir, 'bin');
  mkdirSync(binDir, { recursive: true });
  const binaryPath = join(binDir, BINARY);

  if (!artifactsDir) {
    writeExecutable(binaryPath, "#!/usr/bin/env sh\necho 'placeholder binary for npm package dry-run validation'\n");
  } else {
    extractBinary(findArchive(artifactsDir, target), binaryPath);
  }

  writeJson(join(packageDir, 'package.json'), {
    name: meta.package,
    version,
    description: `Heimdall sandbox CLI binary for ${target}.`,
    license: 'MIT',
    repository: { type: 'git', url: 'git+https://github.com/casualjim/heimdall-sandbox.git' },
    homepage: 'https://github.com/casualjim/heimdall-sandbox',
    os: [meta.os],
    cpu: [meta.cpu],
    bin: { [BINARY]: `bin/${BINARY}` },
    files: ['bin', 'README.md'],
  });
  writeFileSync(join(packageDir, 'README.md'), `# ${meta.package}\n\nPlatform binary package for \`${PACKAGE}\` on \`${target}\`.\n`, 'utf8');
  return packageDir;
}

function createMainPackage(outDir: string, version: string): string {
  const packageDir = join(outDir, 'heimdall-sandbox');
  const binDir = join(packageDir, 'bin');
  mkdirSync(binDir, { recursive: true });
  const optionalDependencies = Object.fromEntries(Object.values(TARGETS).map((meta) => [meta.package, version]));

  writeJson(join(packageDir, 'package.json'), {
    name: PACKAGE,
    version,
    description: 'Process sandbox runtime for Heimdall.',
    license: 'MIT',
    repository: { type: 'git', url: 'git+https://github.com/casualjim/heimdall-sandbox.git' },
    homepage: 'https://github.com/casualjim/heimdall-sandbox',
    bin: { [BINARY]: `bin/${BINARY}.js` },
    optionalDependencies,
    files: ['bin', 'README.md'],
  });

  writeExecutable(
    join(binDir, `${BINARY}.js`),
    `#!/usr/bin/env node
const { spawnSync } = require('node:child_process');
const process = require('node:process');

const packages = {
  'linux:x64': '@casualjim/heimdall-sandbox-linux-x64',
  'linux:arm64': '@casualjim/heimdall-sandbox-linux-arm64',
  'darwin:arm64': '@casualjim/heimdall-sandbox-darwin-arm64',
};

const key = \`${'${process.platform}:${process.arch}'}\`;
const packageName = packages[key];
if (!packageName) {
  console.error(\`Unsupported platform ${'${key}'}. Supported platforms: ${'${Object.keys(packages).join(\', \')}'}\`);
  process.exit(1);
}

let binary;
try {
  binary = require.resolve(\`${'${packageName}'}/bin/heimdall-sandbox\`);
} catch (error) {
  console.error(\`Missing Heimdall platform package ${'${packageName}'}. Reinstall @casualjim/heimdall-sandbox and ensure optional dependencies are enabled.\`);
  process.exit(1);
}

const result = spawnSync(binary, process.argv.slice(2), { stdio: 'inherit' });
if (result.error) {
  console.error(\`Failed to execute ${'${binary}'}: ${'${result.error.message}'}\`);
  process.exit(1);
}
process.exit(result.status ?? 1);
`,
  );
  writeFileSync(
    join(packageDir, 'README.md'),
    `# ${PACKAGE}\n\nRegistry-hosted Heimdall CLI package. Platform binaries are supplied by optional npm dependencies; no GitHub release asset download occurs during install or first run.\n`,
    'utf8',
  );
  return packageDir;
}

function npmPackDryRun(packageDir: string): void {
  spawnChecked('npm', ['pack', '--dry-run'], undefined, packageDir);
}

function spawnChecked(command: string, args: string[], output: 'buffer' | undefined = undefined, cwd?: string): { stdout: string; stderr: string } | { stdout: Buffer; stderr: Buffer } {
  const result = spawnSync(command, args, {
    cwd,
    encoding: output === 'buffer' ? undefined : 'utf8',
    stdio: output === 'buffer' ? ['ignore', 'pipe', 'pipe'] : ['ignore', 'pipe', 'inherit'],
  });
  if (result.status !== 0) {
    const stderr = Buffer.isBuffer(result.stderr) ? result.stderr.toString('utf8') : (result.stderr ?? '');
    throw new Error(`${command} ${args.join(' ')} failed: ${stderr}`);
  }
  return { stdout: result.stdout as never, stderr: result.stderr as never };
}

function main(): void {
  const args = parseArguments(process.argv.slice(2));
  const version = args.version ?? workspaceVersion();
  const artifactsDir = args.dryRunPlaceholders ? undefined : args.artifactsDir;

  if (!artifactsDir && !args.dryRunPlaceholders) {
    throw new Error('--artifacts-dir is required unless --dry-run-placeholders is set');
  }
  if (existsSync(args.outDir)) {
    rmSync(args.outDir, { recursive: true, force: true });
  }
  mkdirSync(args.outDir, { recursive: true });

  const packageDirs = Object.entries(TARGETS).map(([target, meta]) => createPlatformPackage(args.outDir, target, meta, version, artifactsDir));
  packageDirs.push(createMainPackage(args.outDir, version));

  if (args.packDryRun) {
    for (const packageDir of packageDirs) {
      npmPackDryRun(packageDir);
    }
  }

  console.log(packageDirs.join('\n'));
}

try {
  main();
} catch (error) {
  console.error(error instanceof Error ? error.message : error);
  process.exit(1);
}

File: scripts/prepare-npm-packages.py
#!/usr/bin/env python3
"""Assemble registry-hosted npm packages from cargo-dist artifacts."""

from __future__ import annotations

import argparse
import json
import os
import shutil
import stat
import subprocess
import re
import tarfile
from pathlib import Path

PACKAGE = "@casualjim/heimdall-sandbox"
BINARY = "heimdall-sandbox"
TARGETS = {
    "x86_64-unknown-linux-gnu": {
        "package": "@casualjim/heimdall-sandbox-linux-x64",
        "os": "linux",
        "cpu": "x64",
    },
    "aarch64-unknown-linux-gnu": {
        "package": "@casualjim/heimdall-sandbox-linux-arm64",
        "os": "linux",
        "cpu": "arm64",
    },
    "aarch64-apple-darwin": {
        "package": "@casualjim/heimdall-sandbox-darwin-arm64",
        "os": "darwin",
        "cpu": "arm64",
    },
}


def workspace_version() -> str:
    manifest = Path("Cargo.toml").read_text(encoding="utf-8")
    in_workspace_package = False
    for line in manifest.splitlines():
        stripped = line.strip()
        if stripped == "[workspace.package]":
            in_workspace_package = True
            continue
        if stripped.startswith("[") and stripped != "[workspace.package]":
            in_workspace_package = False
        if in_workspace_package:
            match = re.match(r'version\s*=\s*"([^"]+)"', stripped)
            if match:
                return match.group(1)
    raise SystemExit("workspace.package.version not found in Cargo.toml")


def write_json(path: Path, data: dict[str, object]) -> None:
    path.write_text(json.dumps(data, indent=2, sort_keys=False) + "\n", encoding="utf-8")


def write_executable(path: Path, content: str) -> None:
    path.write_text(content, encoding="utf-8")
    mode = path.stat().st_mode
    path.chmod(mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)


def find_archive(artifacts_dir: Path, target: str) -> Path:
    matches = sorted(artifacts_dir.glob(f"{BINARY}-{target}.tar.*"))
    if not matches:
        raise SystemExit(f"missing cargo-dist archive for {target} in {artifacts_dir}")
    return matches[0]


def extract_binary(archive: Path, destination: Path) -> None:
    with tarfile.open(archive) as tar:
        for member in tar.getmembers():
            if Path(member.name).name != BINARY or not member.isfile():
                continue
            source = tar.extractfile(member)
            if source is None:
                raise SystemExit(f"archive {archive} could not read {member.name}")
            with destination.open("wb") as output:
                shutil.copyfileobj(source, output)
            destination.chmod(destination.stat().st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
            return
    raise SystemExit(f"archive {archive} does not contain {BINARY}")


def package_slug(package_name: str) -> str:
    return package_name.removeprefix("@casualjim/")


def create_platform_package(out_dir: Path, target: str, meta: dict[str, str], version: str, artifacts_dir: Path | None) -> Path:
    package_dir = out_dir / package_slug(meta["package"])
    bin_dir = package_dir / "bin"
    bin_dir.mkdir(parents=True, exist_ok=True)
    binary_path = bin_dir / BINARY
    if artifacts_dir is None:
        write_executable(binary_path, "#!/usr/bin/env sh\necho 'placeholder binary for npm package dry-run validation'\n")
    else:
        extract_binary(find_archive(artifacts_dir, target), binary_path)

    write_json(
        package_dir / "package.json",
        {
            "name": meta["package"],
            "version": version,
            "description": f"Heimdall sandbox CLI binary for {target}.",
            "license": "MIT",
            "repository": {"type": "git", "url": "git+https://github.com/casualjim/heimdall-sandbox.git"},
            "homepage": "https://github.com/casualjim/heimdall-sandbox",
            "os": [meta["os"]],
            "cpu": [meta["cpu"]],
            "bin": {BINARY: f"bin/{BINARY}"},
            "files": ["bin", "README.md"],
        },
    )
    (package_dir / "README.md").write_text(
        f"# {meta['package']}\n\nPlatform binary package for `{PACKAGE}` on `{target}`.\n",
        encoding="utf-8",
    )
    return package_dir


def create_main_package(out_dir: Path, version: str) -> Path:
    package_dir = out_dir / "heimdall-sandbox"
    bin_dir = package_dir / "bin"
    bin_dir.mkdir(parents=True, exist_ok=True)
    optional_dependencies = {meta["package"]: version for meta in TARGETS.values()}
    write_json(
        package_dir / "package.json",
        {
            "name": PACKAGE,
            "version": version,
            "description": "Process sandbox runtime for Heimdall.",
            "license": "MIT",
            "repository": {"type": "git", "url": "git+https://github.com/casualjim/heimdall-sandbox.git"},
            "homepage": "https://github.com/casualjim/heimdall-sandbox",
            "bin": {BINARY: f"bin/{BINARY}.js"},
            "optionalDependencies": optional_dependencies,
            "files": ["bin", "README.md"],
        },
    )
    write_executable(
        bin_dir / f"{BINARY}.js",
        """#!/usr/bin/env node
const { spawnSync } = require('node:child_process');
const process = require('node:process');

const packages = {
  'linux:x64': '@casualjim/heimdall-sandbox-linux-x64',
  'linux:arm64': '@casualjim/heimdall-sandbox-linux-arm64',
  'darwin:arm64': '@casualjim/heimdall-sandbox-darwin-arm64',
};

const key = `${process.platform}:${process.arch}`;
const packageName = packages[key];
if (!packageName) {
  console.error(`Unsupported platform ${key}. Supported platforms: ${Object.keys(packages).join(', ')}`);
  process.exit(1);
}

let binary;
try {
  binary = require.resolve(`${packageName}/bin/heimdall-sandbox`);
} catch (error) {
  console.error(`Missing Heimdall platform package ${packageName}. Reinstall @casualjim/heimdall-sandbox and ensure optional dependencies are enabled.`);
  process.exit(1);
}

const result = spawnSync(binary, process.argv.slice(2), { stdio: 'inherit' });
if (result.error) {
  console.error(`Failed to execute ${binary}: ${result.error.message}`);
  process.exit(1);
}
process.exit(result.status ?? 1);
""",
    )
    (package_dir / "README.md").write_text(
        f"# {PACKAGE}\n\nRegistry-hosted Heimdall CLI package. Platform binaries are supplied by optional npm dependencies; no GitHub release asset download occurs during install or first run.\n",
        encoding="utf-8",
    )
    return package_dir


def npm_pack_dry_run(package_dir: Path) -> None:
    subprocess.run(["npm", "pack", "--dry-run"], cwd=package_dir, check=True)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--version", default=workspace_version())
    parser.add_argument("--artifacts-dir", type=Path)
    parser.add_argument("--out-dir", type=Path, default=Path("target/npm-packages"))
    parser.add_argument("--dry-run-placeholders", action="store_true")
    parser.add_argument("--pack-dry-run", action="store_true")
    args = parser.parse_args()

    artifacts_dir = None if args.dry_run_placeholders else args.artifacts_dir
    if artifacts_dir is None and not args.dry_run_placeholders:
        raise SystemExit("--artifacts-dir is required unless --dry-run-placeholders is set")

    if args.out_dir.exists():
        shutil.rmtree(args.out_dir)
    args.out_dir.mkdir(parents=True)

    package_dirs = [create_platform_package(args.out_dir, target, meta, args.version, artifacts_dir) for target, meta in TARGETS.items()]
    package_dirs.append(create_main_package(args.out_dir, args.version))

    if args.pack_dry_run:
        for package_dir in package_dirs:
            npm_pack_dry_run(package_dir)

    print(os.linesep.join(str(path) for path in package_dirs))


if __name__ == "__main__":
    main()
